syntax = "proto2";
enum filetype {
  ft_unknown   = 0;
  ft_socket    = 0140000;
  ft_symlink   = 0120000;
  ft_regular   = 0100000;
  ft_block     = 0060000;
  ft_directory = 0040000;
  ft_character = 0020000;
  ft_fifo      = 0010000;
}

message FilesystemMetadata {
  required uint32 version = 1;
  required uint32 blockbits = 2;
}

message Timespec {
  required int64 sec = 1;
  required uint32 nsec = 2;
}

message ProcessTableEntry {
  // randomly generated 128 bits
  required bytes pid = 1;
  // based on the processes' clock. not reliable.
  required Timespec last_updated = 2;
  // incremented every time this is updated
  required uint64 liveness_counter = 3;

  optional string hostname = 4;
}

message StatRecord {
  required fixed64 ino = 1;
  required fixed64 dev = 2;
  required uint32 mode = 3;
  required uint64 nlink = 4;
  required uint32 uid = 5;
  required uint32 gid = 6;
  required int64 size = 7;
  required Timespec atime = 8;
  required Timespec mtime = 9;
  required Timespec ctime = 10;
  required uint32 blksize = 11;
  required uint64 blocks = 12;
}

message OpLogResultOK {}

message OpLogResultEntry {
  required fixed64 ino = 1;
  required uint64 generation = 2;
  required StatRecord attr = 3;
  required double attr_timeout = 4;
  required double entry_timeout = 5;
}

message OpLogResultAttr {
  required StatRecord attr = 1;
}

message OpLogResultOpen {
  required fixed64 ino = 1;
  required uint32 flags = 2;
  required bool direct_io = 3;
  required bool keep_cache = 4;
  required bool nonseekable = 5;
}

message OpLogResultBuf {
  required bytes data = 1;
  optional uint64 actual_size = 2;
}

message OpLogResultReadlink {
  required bytes target = 1;
}

message OpLogResultWrite {
  required uint64 size = 1;
}

message OpLogResultStatfs {
  required uint64 bsize = 1;
  required uint64 frsize = 2;
  required uint64 blocks = 3;
  required uint64 bfree = 4;
  required uint64 bavail = 5;
  required uint64 files = 6;
  required uint64 ffree = 7;
  required uint64 favail = 8;
  required uint64 fsid = 9;
  required uint64 flag = 10;
  required uint64 namemax = 11;
}

message OpLogResultXattrSize {
  required int64 size = 1;
}

message OpLogRecord {
  // Monotonic per-process operation id.
  required uint64 op_id = 1;
  // Structured operation result payload for replay.
  oneof result {
    OpLogResultOK ok = 10;
    OpLogResultEntry entry = 11;
    OpLogResultAttr attr = 12;
    OpLogResultOpen open = 13;
    OpLogResultBuf buf = 14;
    OpLogResultReadlink readlink = 15;
    OpLogResultWrite write = 16;
    OpLogResultStatfs statfs = 17;
    OpLogResultXattrSize xattr_size = 18;
  }
}

message DirectoryEntry {
  // We can plunk anything in here from the
  // INodeRecord that's immutable
  required fixed64 inode = 1;
  required filetype type = 2;
  // symlink value?
}

message INodeRecord {
  // 0.. fundamentals
  required fixed64 inode = 1; // IMMUTABLE
  required filetype type = 2; // IMMUTABLE
  required uint32 nlinks = 3;
  optional uint32 mode = 4;
  optional uint64 size = 5;
  optional uint32 rdev = 6;

  // 100... ownership block
  optional uint32 uid = 100;
  optional uint32 gid = 101;

  // 200... time block
  optional Timespec atime = 200;
  optional Timespec mtime = 201;
  optional Timespec ctime = 202;

  // 1000... unix weirdness
  optional string symlink = 1000;
}

message XAttrRecord {
  // 0.. fundamentals

  // why do we have a message to store nothing?
  // we might want to add permissions, or per-xattr metadata,
  // or keep the size in the node, etc
}
